### PR TITLE
Introduce DEAL_II_IF_CONSTEXPR

### PR DESCRIPTION
**Pull request by masterleinad**
https://github.com/dealii/dealii/pull/14571 will introduce a bunch of places where I need `if constexpr` at least for compiling with `Kokkos+SYCL` (which requires C++17 anyway) but other backends (like `Cuda`, `HIP`, `OpenMP`, or `Serial`) don't need it. Hence, there are a bunch of places that look like
```C++
#ifdef DEAL_II_HAVE_CXX17
  if constexpr (condition)
#else
  if (condition)
#endif
```
and we have similar code already in the code base where we found using `if constexpr` helpful already. Of course, in any of these cases, the guarded code must still fully compile limiting the usefulness of actually doing this.
This pull request proposes to introduce a macro that simplifies writing these code blocks to enhance readability.
Unfortunately, `clang-format` only supports `IfMacros` from clang-format 13 and `ForEachMacros` seemed to be the next best thing. That's why there are a lot of additional white space changes.

### CODE REVIEW

File: include/deal.II/base/config.h.in [hunk 1]
```diff
@@ -545,6 +545,19 @@ _Pragma("GCC diagnostic pop")
...
+ * Some places in deal.II require or benefit from using "if constexpr".
+ * However, this feature is only available in C++17 and above.
+ * The macro defined here expands to "if constexpr(...)" when C++17 is supported
+ * and to "if (...)" otherwise.
+ */
+#ifdef DEAL_II_HAVE_CXX17
+  #define DEAL_II_IF_CONSTEXPR(x) if constexpr x
+#else
+  #define DEAL_II_IF_CONSTEXPR(x) if x
+#endif
```
Comment (drwells) [inline include/deal.II/base/config.h.in:558]: What's the advantage of `if constexpr x` over `if constexpr (x)`?

Comment (masterleinad) [reply include/deal.II/base/config.h.in:558]: For anything including a comma we need to use double parentheses anyway and this forces double parentheses for every condition. I thought that would be good for consistency but I'm happy to change.

**Comment by tjhei** [comment]
Just an idea: if you leave the "if" outside the macro and only include "constexpr x", would that fix the indentation?

**Comment by bangerth** [comment]
I'm with @tjhei : I think this would solve most of the problems:
```
  if DEAL_II_CONSTEXPR_IN_CONDITIONAL (x)
```
where `DEAL_II_CONSTEXPR_IN_CONDITIONAL` is either an empty preprocessor define, or `constexpr` in C++17.

**Comment by bangerth** [comment]
By the way, this will also allow you to use the same in `switch consexpr (x)` places if necessary.

**Comment by masterleinad** [comment]
> By the way, this will also allow you to use the same in `switch consexpr (x)` places if necessary.

Unfortunately, you can't use qconstexpr` with `switch`. Otherwise, I followed your suggestion now.

**Review by bangerth — APPROVED** [review]
This looks good to me now, thank you!

**Pull request merged**